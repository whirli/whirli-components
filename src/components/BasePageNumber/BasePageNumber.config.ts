// @ts-ignore
import { PropType } from 'vue';
import * as PropHelpers from '../../helpers/props';
import { ComponentStyles, ComponentProps } from '../../@types/components';

// Local config
import config from '@whirli-local/components/BasePageNumber/BasePageNumber.config';

// Package config
import { PropKeys } from './BasePageNumber.constants';

export const ConfigStyles: ComponentStyles = {
  ...config.styles,
};

export const ConfigProps: ComponentProps = {
  [PropKeys.PAGE_NUMBER]: {
    required: true,
    type: Number as PropType<number>,
  },
  ...config.props,
};
