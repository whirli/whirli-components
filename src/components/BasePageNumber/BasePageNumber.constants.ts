export const PropKeys: Record<string, string> = {
  PAGE_NUMBER: 'pageNumber',
  IS_DISABLED: 'isDisabled',
};

export const PropValues: Record<string, string[]> = {};
