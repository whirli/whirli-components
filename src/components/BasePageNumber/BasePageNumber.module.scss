@import '@whirli-components/styles/Breakpoints.module.scss';
@import '@whirli-components/styles/Spacers.scss';

// default styles
.page-number {
  display: flex;
  height: $spacer-md;
  width: $spacer-md;
  justify-content: center;
  align-items: center;
  text-decoration: none;
  margin: 0 $spacer-3xs;
}

.is-disabled {
  cursor: not-allowed;
  opacity: 0.5;
}

.is-active { background-color: #808080;}

// breakpoint styles
$package-breakpoint-styles: (

);
