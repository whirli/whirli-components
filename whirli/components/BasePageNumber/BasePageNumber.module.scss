@import '@whirli-components/components/BasePageNumber/BasePageNumber.module.scss';

// Standard styles

// Breakpoint styles
$localBreakpointStyles: ();

@include createBreakpointClasses(map-merge($packageBreakpointStyles, $localBreakpointStyles));