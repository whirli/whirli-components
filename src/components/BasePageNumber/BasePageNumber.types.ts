import { PropKeys, PropValues } from './BasePageNumber.constants';
