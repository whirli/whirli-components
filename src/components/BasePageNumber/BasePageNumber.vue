<template>
  <component
    :is="tag"
    :href="pageUrl"
    :class="[defaultClasses, activeClasses, disabledClasses]"
    @click.prevent="$emit('page-number:click', props.pageNumber)"
  >
    {{ props.pageNumber }}
  </component>
</template>

<script setup lang="ts">
// Vue
import { defineEmits, computed } from '@composition';
import type { ComputedRef } from '@composition';
import { useRoute } from 'vue-router';

defineEmits(['page-number:click']);
const Route = useRoute();

// Styles
// @ts-ignore
import styles from '@whirli-local/components/BasePageNumber/BasePageNumber.module.scss?module';

// Data
import { ConfigStyles, ConfigProps } from './BasePageNumber.config';
import { PropKeys } from '@whirli-components/components/BasePageNumber/BasePageNumber.constants';

// Types
import { ComponentStyles as ComponentStylesInterface } from '@whirli-components/@types/components';
import { Props } from '@whirli-components/@types/props';

const ComponentStyles: ComponentStylesInterface = ConfigStyles;

// @ts-ignore
const props: Props = defineProps(ConfigProps);

const pageUrl: ComputedRef<string> = computed(() => `${Route.path}?page=${props.pageNumber}`);
const isActivePage: ComputedRef<boolean> = computed(
  () =>
    (Route.query.page && +Route.query.page === props.pageNumber) ||
    (!Route.query.page && props.pageNumber === 1)
);

const tag: ComputedRef<string> = computed(() => (isActivePage.value ? 'div' : 'a'));

// Classes
import useClasses from '@whirli-components/@use/class';
const { makeClasses } = useClasses();
const defaultClasses = [styles['page-number'], ...makeClasses(ComponentStyles, props, styles)];
const activeClasses: ComputedRef<string> = computed(() => (isActivePage.value ? styles['is-active'] : ''));
const disabledClasses: ComputedRef<string> = computed(() =>
  props[PropKeys.IS_DISABLED] ? styles['is-disabled'] : ''
);
</script>
