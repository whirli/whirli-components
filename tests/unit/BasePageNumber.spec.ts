import { mount } from '@vue/test-utils';
import BasePageNumber from '@/components/BasePageNumber/BasePageNumber.vue';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
let wrapper: any;
const pageNumber = '2';

jest.mock('vue-router', () => ({
  useRoute: () => ({
    query: {
      page: '3',
    },
  }),
}));

describe('BaseFormLabel.vue', () => {
  beforeEach(() => {
    wrapper = mount(BasePageNumber, {
      props: {
        pageNumber,
        tag: 'a',
      },
    });
  });

  it('renders the component', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('displays the correct text', () => {
    expect(wrapper.text()).toMatch(pageNumber);
  });

  it('has the correct tag', () => {
    expect(wrapper.element.tagName).toBe('A');
  });

  it('emmits page number on click', async () => {
    wrapper.find('a').trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('page-number:click'));
  });
});
