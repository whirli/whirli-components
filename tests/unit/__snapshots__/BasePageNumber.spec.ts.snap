// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseFormLabel.vue renders the component 1`] = `
<a
  class=""
  href="undefined?page=2"
  tag="a"
>
  2
</a>
`;
